active-rfcs/0000-finalize-head-body.md

```diff
@@ -0,0 +1,80 @@
+
+
+# Detailed design
+
+## Template Changes
+
+- `<!DOCTYPE html>`
+  - compiler: completely ignored by the compiler, stripped out from render output.
+  - runtime: a `<!DOCTYPE html>` tag is always included in final page HTML output.
+- `<html>`, `<body>`, `<head>`
+  - compiler: will be left as-is in the component template.
+  - runtime: will be left as-is in final page HTML output.
+  - runtime: may warn if duplicate `html`, `body`, and `head` tags are included in final page HTML output.
+
+## Compiler Changes
+
+- `src/pages` and `src/layouts` will no longer be parsed or rendered differently from other components
+- `as?: 'document' | 'fragment';` will be removed from the compiler.
+- All Astro components will now be compiled as fragments.
+
+Losing `"as": "document"` parse mode will remove some special parser handling, like getting an implicit `<head>` to wrap a standalone `<title>` element. This is intentional to bring us more inline with the RFC, where we respect the HTML authored by the developer as the source of truth and leave mostly as-is in the final output.
+
+In this design it is more "on you" to write valid HTML. This comes from the reality that an imported component can contain unknown HTML, so the compiler can't implicitly assume anything about what is or is not included included the final parent component template.  See https://github.com/withastro/astro/issues/2022 for examples of when this breaks down today. We can help with some static linting, and runtime warnings if the final HTML output is invalid. However, this RFC acknowledges the reality that already exists in v0.21 that imported components break any assumptions and help that we previously attempted to provide.
+
+# Drawbacks
+
+- This RFC will allow you to author a page of HTML that does not actually output `<html>`, `<body>` or `<head>` elements. This can be considered an advantage in that it is more flexible than something like Svelte or Vue. However, it means we need to be more diligent with testing this kind of output across our codebase and dependencies. For example, we would need to confirm that Vite does not have trouble injecting things into an HTML document without a head or body.
+  - If we are blocked by this limitation in some way and can't reasonably unblock (for example, a non-trivial Vite limitaiton) this RFC states that it would be acceptable for Astro to add `<head>` or `<body>` elements in a way that keeps the output HTML valid but unblocks Astro. This would not be done unless absolutely necessary. 
+
+# Alternatives
+
+- Forcing that a user includes `<html>`, `<body>` and `<head>` elements in the final page output. This could be explored more, however there is not currently consensus around this idea and also there is no known reason to tackle this as a part of this RFC. If this is desired, it should be considered a seperate feature request and out of scope from this RFC.
```

Comment:
I think for SSR it makes sense that we require there be a `<head>` in *some* component (could be any component). This is because we will likely need to inject things into the head, and the best way to do that is for the compiler to tell us where the end of the head is. But we cannot know that without there being an ending head tag within some component.

Happy to explain this detail offline if this description doesn't explain it well enough.

Reply:
I think that makes sense. If no head exists, is it safe to inject into the top of the `<html>` element instead?

Reply:
If we’re omitting `<!doctype html>`, pretty sure the compiler could inject everything at the top, and it would be considered `<head>`.

Intelligently finding the first `<head>` below some HTML5-Boilerplate-like comments that may exist in a `<Layout>` seems much more complicated.

Reply:
An aside about `<head>` —

The browser will generate `<head>` in any current HTML document. The DOM has a built-in shorthand to get it (`document.head`). `<head>` has no role, no attributes, and — unlike `<html>` or `<body>` — there have been no popular inventions that required adding classes or IDs to it.

Crawlers, open-graph see exciting things all look for individual `<meta>` & `<title>` tags anyway.

My two cents: Don’t fuss with `<head>`. The web 'has your back'.

Reply:
The browser doesn't need head but Astro does. Astro injects things into the head, in particular it injects component styles. So we need to know where the head is. Given an `index.astro` that looks like this:

```astro
<One />
<Two />
<Three />
```

The head could be in One, or One and Two, or One, Two and Three. Any of these components could be a framework component, so we can't inspect them to know if they include head content or not (aside from parsing the generated HTML, which we don't want to do).

Open to suggestions but I just don't see a way to implement head injections without there being a head in one .astro component.

Reply:
We are okay injecting into the generated HTML though, correct? As a pass on the final page HTML output by finding the closing `</head>` element? If this happens at runtime, then I don't think that requires any sort of ahead-of-time analysis.

https://github.com/withastro/rfcs/blob/FredKSchott-patch-2/active-rfcs/0000-finalize-head-body.md#drawbacks The drawback section addresses this more directly, especially the  "If we are blocked by this limitation in some way and can't reasonably unblock" point.

The assumption is that we could inject into the top of the page if no `<head />` exists, but I agree that there's a strong chance our tooling & Vite wouldn't support this (Vite tries to inject things into head as well), in which case requiring a head (with a runtime error if none is found) becomes reasonable. I'll update the RFC to make this more clear.

Reply:
Yes, we inject it post-render by parsing the HTML. In an SSR scenario this seems too costly to continue to do. It also prevents streaming.

Reply:
Re:
> The head could be in One, or One and Two, or One, Two and Three. Any of these components could be a framework component, so we can't inspect them to know if they include head content or not (aside from parsing the generated HTML, which we don't want to do).

If we can’t inspect those components, then wouldn’t that require at least every `.astro` page to define the `<html lang>` and `<head>` without any conditional surrounding conditional logic?

Re:
> `src/pages` and `src/layouts` will no longer be parsed or rendered differently from other components

If we can’t inspect components, then is this compiler change no longer possible?

Reply:
Maybe this is a change to how we render page HTML then? I don't think there's anything blocking our ability to, when rendering the page, say "ah yes, the head element, great. I've been looking for you. I'll just inject some things here before outputting the final page HTML."

@jonathantneal if you're still -1 on this, lets all talk on Monday. I think we've made our best effort here to support no-`<head>` HTML authoring development. But since we need to inject stuff into the head, I agree with Matthew that that would be significantly more difficult without an actual `<head>` component to inject into.

Happy to talk through alternatives but I think we could benefit from a quicker back-and-forth given the complexity here.

Reply:
I want to offer some peace of mind that I’m +1 on injection if it’s needed for SSR performance.

> I don't think there's anything blocking our ability to, when rendering the page, say "ah yes, the head element, great. I've been looking for you..."

This is the bit I’d like us to discuss Monday. It definitely aligns with what you mentioned earlier about finding the closing `</head>` element (presumably from the page _or_ a component), but that doesn’t align with what @matthewp wrote about a need to not inspect components to know if they include head content or not (specifically to not parse the generated HTML).

Reply:
Yup, I'd like to understand this better as well. 

Oddly enough, you've gotten me excited now about the idea of supporting `<head>`-less HTML authoring -- as long as its not a non-trivial amount of extra work. I feel like I started out this process without caring much about it, and now I'm almost sad at the idea of losing it!

Reply:
My proposal would only require that 1 astro component includes the `<head>` tag. It could be any astro component; pages, layouts, regular components.

My opinion is that rendering should happen like this `render() -> final HTML -> browser`. Currently it works like this `render() -> rendered HTML -> parse -> injections -> finalHTML -> browser`. I think we need a very good reason to continue doing it like this.

Reply:
Talked in person, and are all onboard with requiring a single `<head>` element for head injection. RFC updated to include latest thinking and additional details.